pkg/ctrl/fuse.go

```diff
@@ -27,82 +27,89 @@ import (
 	"github.com/pkg/errors"
-	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/labels"
-	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // CheckFuseHealthy checks the ds healthy with role
-func (e *Helper) CheckFuseHealthy(recorder record.EventRecorder, runtime base.RuntimeInterface,
-	currentStatus datav1alpha1.RuntimeStatus,
-	ds *appsv1.DaemonSet) (err error) {
-	var (
-		healthy             bool
-		unavailablePodNames []types.NamespacedName
-	)
-	if ds.Status.NumberUnavailable == 0 {
-		healthy = true
+func (e *Helper) CheckFuseHealthy(recorder record.EventRecorder, runtime base.RuntimeInterface, fuseName string) error {
+	currentStatus := runtime.GetStatus()
+
+	runtimeInfo := e.runtimeInfo
+	ds, err := kubeclient.GetDaemonset(e.client, fuseName, runtimeInfo.GetNamespace())
+	if err != nil {
+		return err
 	}
 
-	statusToUpdate := runtime.GetStatus()
-	if len(statusToUpdate.Conditions) == 0 {
-		statusToUpdate.Conditions = []datav1alpha1.RuntimeCondition{}
+	healthy := true
+	if ds.Status.NumberUnavailable > 0 ||
+		(ds.Status.DesiredNumberScheduled > 0 && ds.Status.NumberAvailable == 0) {
+		healthy = false
 	}
 
-	if healthy {
-		cond := utils.NewRuntimeCondition(datav1alpha1.RuntimeFusesReady, "The fuse is ready.",
-			"The fuse is ready.", corev1.ConditionTrue)
-		_, oldCond := utils.GetRuntimeCondition(statusToUpdate.Conditions, cond.Type)
+	updateRuntimeStatus := func(runtimePhase datav1alpha1.RuntimePhase, conditionStatus corev1.ConditionStatus,
+		reason, message string) error {
+		statusToUpdate := runtime.GetStatus()
 
-		if oldCond == nil || oldCond.Type != cond.Type {
-			statusToUpdate.Conditions =
-				utils.UpdateRuntimeCondition(statusToUpdate.Conditions,
-					cond)
-		}
-		statusToUpdate.FusePhase = datav1alpha1.RuntimePhaseReady
-	} else {
-		// 1. Update the status
-		cond := utils.NewRuntimeCondition(datav1alpha1.RuntimeFusesReady, "The fuses are not ready.",
-			fmt.Sprintf("The fuses %s in %s are not ready.", ds.Name, ds.Namespace), corev1.ConditionFalse)
+		cond := utils.NewRuntimeCondition(datav1alpha1.RuntimeFusesReady, reason, message, conditionStatus)
 		_, oldCond := utils.GetRuntimeCondition(statusToUpdate.Conditions, cond.Type)
 
 		if oldCond == nil || oldCond.Type != cond.Type {
 			statusToUpdate.Conditions =
 				utils.UpdateRuntimeCondition(statusToUpdate.Conditions,
 					cond)
 		}
-		statusToUpdate.FusePhase = datav1alpha1.RuntimePhaseNotReady
+		statusToUpdate.FusePhase = runtimePhase
+		statusToUpdate.FuseNumberReady = int32(ds.Status.NumberReady)
+		statusToUpdate.FuseNumberAvailable = int32(ds.Status.NumberAvailable)
+		if !reflect.DeepEqual(*statusToUpdate, currentStatus) {
+			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+				return e.client.Status().Update(context.TODO(), runtime)
```

Comment:
I'm concerned that `RetryOnConflict` is used without incorporating the latest changes, it may repeatedly fail due to persistent conflicts.